src/OrchardCore.Modules/OrchardCore.BackgroundTasks/Services/BackgroundTaskDocumentSettingsProvider.cs

```diff
@@ -0,0 +1,38 @@
+using System;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace OrchardCore.BackgroundTasks.Services
+{
+    public class BackgroundTaskDocumentSettingsProvider : IBackgroundTaskSettingsProvider
+    {
+        private readonly IHttpContextAccessor _httpContextAccessor;
+
+        public BackgroundTaskDocumentSettingsProvider(IHttpContextAccessor httpContextAccessor)
```

Comment:
Hum, this doesn't look right.

I understand that you can't inject `BackgroundTaskManager` (this class is used in `BackgroundTaskDocumentOptionsSetup`, which has a longer lifetime than `BackgroundTaskManager`), but I doubt using the context accessor is the correct solution. Instead, why not injecting `IServiceProvider` and creating an ad-hoc scope in `GetSettingsAsync()`?

```csharp
using (var scope = _serviceProvider.CreateScope())
{
    var manager = scope.ServiceProvider.GetRequiredService<BackgroundTaskManager>();
}
```

Reply:
Yes, this is to retrieve the service provider of the current scope which is our scope wrapper which updates in its ctor the `HttpContext.RequestServices`.

Yes, we could do this but, because `BackgroundTaskManager` inject an `ISession`, doing so will create a new session.

We use this provider before invoking a bgTask and if you have e.g 3 tasks i can see that 3 sessions are created. Then, when creating a new scope in this provider, i can see that 6 sessions are created.

Reply:
@jtkech if creating an `ISession` is too expensive, one solution is to inject `IServiceProvider` instead of `ISession` in `BackgroundTaskManager` and only resolve `ISession` when you need it (e.g when the document is not already cached).

Reply:
Just tried and it works, good idea but at the moment i'm a little lost, i always need to re-put all of this into my mind ;) I think we need to discuss on general rules around `IHttpContextAccessor`, child scopes,  and fake `HttpContext`, meanwhile, as i remember, some general considerations.

- Hmm, so here we would depend on how a service injects and / or uses a dependency. And in a class, `IServiceProvider` could be related to the current scope or its `root`.

**How to create a child scope**

- Just as a reminder for me when working in the background (so not in a request scope). Ok to use `_serviceProvider.CreateScope()` if we already are in a tracked scope wrapper.

- Maybe better to always work with a scope synchronized with `.RequestServices` if a service uses it. So, maybe better to always use `EnterServiceScope()` which allows to retrieve / pass the `ServiceProvider` of the current scope through `IHttpContextAccessor`, and a having fake `HttpContext` in the background. But maybe not when we know what is involved, if we limit the usage of `IHttpContextAccessor` ... Or create a custom accessor to retrieve a service provider but not tied to `HttpContext`, but maybe not worth to replicate what is already done.

- As i remember the 1st issue we had was because of using `SiteService` in the background. But maybe `SiteService` could be a scoped service which could resolves the `ISession` without using `IHttpContextAccessor`. Just tried, 1st test seems to work correctly.

- Maybe by doing this for some other services we could get rid of fake `HttpContext`. And then also be more relax in some places to just use `Createscope()` in place of `EnterServiceScope()`.

- Hmm, but i think that anyway we will need to run in the background some services that require an `HttpContext`. If so, maybe what i began here to create a fake `HttpContext` could be embedded in `ServiceScopeWrapper` which right now just create a new `DefaultHttpContext()`.

**About creating or not a child scope**

- With sqlite, if a parent scope update the same table as a nested child, the nested child, waiting for its parent to dispose, is locked. I thought it is a bad limitation, but @sebastienros said that it forces us to do things correctly.

- So, in some cases, if we are in a scope which has already updated a table, **better to not create a child scope** which might update the same table. But in some cases **better to create a child scope** to update a table if we know that afterwards another child scope could do the same.

- Note: we had a db locking issue (only with sqlite) when a recipe includes a step which is a recipe itself, we needed to execute it later, after disposing the scope of the parent recipe.

So, not always sure what is the best, create or not a child scope and how, use or not `IHttpContextAccessor` ... Maybe good to open an issue to find the best compromises. It was not the case before but now, when i see `IHttpContextAccessor` i don't fell me good ;)